### PR TITLE
If enabled, mark pushed time after an event is sent in export distro

### PR DESCRIPTION
**Pull request by davidmsh**
As with java version, mark pushed time after an event is sent in export distro
*1 Add MarkPushed in export distro config to optionally enable this feature
*2 Add MarkPushed() function in coredata client to invoke corresponding rest api
*3 Invoke MarkPushed() at the end of processEvent() if configured to do so

### CODE REVIEW

**Comment by tsconn23** [comment]
Is this a PR targeting Issue #316 which was closed some time ago? It looks like it. Please see my comments there. I stand by them. This PR is on hold.

**Comment by jduranf** [comment]
@davidmsh this implementation doesn't take in account if the event have been sent or not. Maybe the send interface should be modified to something like this:
```
type Sender interface {
	Send(data []byte, event *models.Event) bool
}
```
And then in registration.go do something like this:
```
res := reg.sender.Send(encrypted, event)
if configuration.MarkPushed && res {
    ...
}
```
Anyway, first should be fine to know if this PR will be accepted or not. And then do this modification in this PR or in another one.

**Comment by davidmsh** [comment]
@tsconn23 This PR is targeting Issue #316 but related to Issue #315 in the way pushed timestamp is updated in an event. This PR mark an event's push ts using existing rest api from core-data. 
Regarding Issue #315, we all agree that the content of an event should be immutable but it's valuable to update some of its metadata to reflect its current state (and we store them in the same event) and I believe edgex api doesn't allow event content modification. This PR comes from a PoC from the field which should be valuable to the project. Your thought ?

**Comment by davidmsh** [comment]
@jduranf that's a good idea, let's do that after this PR gets accepted.

**Comment by cloudxxx8** [comment]
@tsconn23, this PR is totally different from #316 .  #316 is the bug of an existing API in core-data, and #315  is the bug of export-distro doesn't follow the original design to notify core-data when the event is pushed.  This PR is to fix #315 only.

**Comment by tsconn23** [comment]
@jpwhitemn Would like us to pull this in. @davidmsh Please rebase the branch to get rid of the out-of-date message. Meanwhile, will review.

File: export/distro/registrations.go [hunk 1]
```diff
@@ -158,6 +158,17 @@ func (reg registrationInfo) processEvent(event *models.Event) {
 	}
 
 	reg.sender.Send(encrypted, event)
+
+	if configuration.MarkPushed {
+		go func(id string) {
```
Comment (tsconn23) [inline export/distro/registrations.go]: Why not utilize a goroutine on line 183 of registrations.go instead?
```
case event := <-reg.chEvent:
reg.processEvent(event)
```
Pass event in as a variable to an anonymous function invoked via goroutine....?

**Comment by tsconn23** [comment]
@drasko @janko-isidorovic @feclare It sounds like from talking to @jpwhitemn that this is a bug and will prevent events from being cleared out of the database based on the "pushed" timestamp. If that's so and if we can get this PR merged, we will need this fix for the 0.6.1 update at the end of June. Can I please get one of you gentlemen to oversee getting this into the California branch for inclusion in 0.6.1?

**Comment by feclare** [comment]
There is one goroutine for each registration. If you have several registrations running this code will call MarkPushed several times(one for each registration) for each event.

A better approach is to have a new goroutine accepting ids in a channel and calling to core-data markpush once for each id. If it have an small id cache (just several seconds) of the pushed id we can only make the call once for id.

**Pull request closed without merging**